File: banan-stats/internal/store/store.go
package store

import (
	"context"
	"database/sql"

	_ "github.com/duckdb/duckdb-go/v2"

	"github.com/khaled/banan-stats/banan-stats/internal/analyzer"
)

type Store struct {
	db *sql.DB
}

func Open(dbPath string) (*Store, error) {
	db, err := sql.Open("duckdb", dbPath)
	if err != nil {
		return nil, err
	}
	db.SetMaxOpenConns(1)
	db.SetMaxIdleConns(1)
	if err := initSchema(db); err != nil {
		_ = db.Close()
		return nil, err
	}
	return &Store{db: db}, nil
}

func (s *Store) Close() error {
	return s.db.Close()
}

func (s *Store) DB() *sql.DB {
	return s.db
}

func initSchema(db *sql.DB) error {
	stmts := []string{
		"CREATE TYPE IF NOT EXISTS agent_type_t AS ENUM ('feed', 'bot', 'browser')",
		"CREATE TYPE IF NOT EXISTS agent_os_t AS ENUM ('Android', 'Windows', 'iOS', 'macOS', 'Linux')",
		`CREATE TABLE IF NOT EXISTS stats (
			date       DATE,
			time       TIME,
			host       VARCHAR,
			path       VARCHAR,
			query      VARCHAR,
			ip         VARCHAR,
			user_agent VARCHAR,
			referrer   VARCHAR,
			type       agent_type_t,
			agent      VARCHAR,
			os         agent_os_t,
			ref_domain VARCHAR,
			mult       INTEGER,
			set_cookie UUID,
			uniq       UUID
		)`,
		"ALTER TABLE stats ADD COLUMN IF NOT EXISTS host VARCHAR",
		"CREATE INDEX IF NOT EXISTS idx_stats_host_date ON stats(host, date)",
	}
	for _, stmt := range stmts {
		if _, err := db.Exec(stmt); err != nil {
			return err
		}
	}
	return nil
}

func (s *Store) Insert(ctx context.Context, lines []analyzer.Line) error {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}

	insertSQL := `INSERT INTO stats
		(date, time, host, path, query, ip, user_agent, referrer, type, agent, os, ref_domain, mult, set_cookie, uniq)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
	stmt, err := tx.PrepareContext(ctx, insertSQL)
	if err != nil {
		_ = tx.Rollback()
		return err
	}
	defer stmt.Close()

	updateSQL := "UPDATE stats SET uniq = ? WHERE set_cookie = ?"
	updStmt, err := tx.PrepareContext(ctx, updateSQL)
	if err != nil {
		_ = tx.Rollback()
		return err
	}
	defer updStmt.Close()

	for _, line := range lines {
		analyzer.Analyze(&line)
		_, err = stmt.ExecContext(
			ctx,
			nullString(line.Date),
			nullString(line.Time),
			nullString(line.Host),
			nullString(line.Path),
			nullString(line.Query),
			nullString(line.IP),
			nullString(line.UserAgent),
			nullString(line.Referrer),
			nullString(line.Type),
			nullString(line.Agent),
			nullString(line.OS),
			nullString(line.RefDomain),
			line.Mult,
			nullUUID(line.SetCookie),
			nullUUID(line.Uniq),
		)
		if err != nil {
			_ = tx.Rollback()
			return err
		}

		if line.SecondVisit && line.Uniq != "" {
			if _, err := updStmt.ExecContext(ctx, line.Uniq, line.Uniq); err != nil {
				_ = tx.Rollback()
				return err
			}
		}
	}

	return tx.Commit()
}

func nullString(s string) interface{} {
	if s == "" {
		return nil
	}
	return s
}

func nullUUID(s string) interface{} {
	if s == "" {
		return nil
	}
	return s
}

File: banan-stats-rs/src/store.rs
use crate::analyzer::{self, Line};
use anyhow::Context;
use duckdb::{params, Connection};
use std::sync::{Arc, Mutex};

pub struct Store {
    conn: Arc<Mutex<Connection>>,
}

impl Store {
    pub fn open(path: &str) -> Result<Self, anyhow::Error> {
        let conn = Connection::open(path).with_context(|| format!("open db {}", path))?;
        for stmt in [
            "CREATE TYPE agent_type_t AS ENUM ('feed', 'bot', 'browser')",
            "CREATE TYPE agent_os_t AS ENUM ('Android', 'Windows', 'iOS', 'macOS', 'Linux')",
        ] {
            if let Err(err) = conn.execute(stmt, []) {
                if !is_existing_type_error(&err) {
                    return Err(err.into());
                }
            }
        }

        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS stats (
                 date       DATE,
                 time       TIME,
                 host       VARCHAR,
                 path       VARCHAR,
                 query      VARCHAR,
                 ip         VARCHAR,
                 user_agent VARCHAR,
                 referrer   VARCHAR,
                 type       agent_type_t,
                 agent      VARCHAR,
                 os         agent_os_t,
                 ref_domain VARCHAR,
                 mult       INTEGER,
                 set_cookie UUID,
                 uniq       UUID
             );
             ALTER TABLE stats ADD COLUMN IF NOT EXISTS host VARCHAR;
             CREATE INDEX IF NOT EXISTS idx_stats_host_date ON stats(host, date);",
        )?;

        Ok(Self {
            conn: Arc::new(Mutex::new(conn)),
        })
    }

    pub async fn insert(&self, lines: Vec<Line>) -> Result<(), anyhow::Error> {
        let conn = self.conn.clone();
        tokio::task::spawn_blocking(move || -> Result<(), anyhow::Error> {
            let mut conn = conn.lock().expect("db lock");
            let tx = conn.transaction()?;

            let mut stmt = tx.prepare(
                "INSERT INTO stats
                 (date, time, host, path, query, ip, user_agent, referrer, type, agent, os, ref_domain, mult, set_cookie, uniq)
                 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            )?;
            let mut upd_stmt = tx.prepare("UPDATE stats SET uniq = ? WHERE set_cookie = ?")?;

            for mut line in lines {
                analyzer::analyze(&mut line);
                stmt.execute(params![
                    null_str(&line.date),
                    null_str(&line.time),
                    null_str(&line.host),
                    null_str(&line.path),
                    null_str(&line.query),
                    null_str(&line.ip),
                    null_str(&line.user_agent),
                    null_str(&line.referrer),
                    null_str(&line.r#type),
                    null_str(&line.agent),
                    null_str(&line.os),
                    null_str(&line.ref_domain),
                    line.mult,
                    null_str(&line.set_cookie),
                    null_str(&line.uniq),
                ])?;

                if line.second_visit && !line.uniq.is_empty() {
                    upd_stmt.execute(params![line.uniq, line.uniq])?;
                }
            }

            tx.commit()?;
            Ok(())
        })
        .await??;
        Ok(())
    }

    pub async fn with_conn<T, F>(&self, func: F) -> Result<T, anyhow::Error>
    where
        T: Send + 'static,
        F: FnOnce(&Connection) -> Result<T, anyhow::Error> + Send + 'static,
    {
        let conn = self.conn.clone();
        tokio::task::spawn_blocking(move || {
            let conn = conn.lock().expect("db lock");
            func(&conn)
        })
        .await?
    }
}

fn null_str(s: &str) -> Option<&str> {
    if s.is_empty() {
        None
    } else {
        Some(s)
    }
}

fn is_existing_type_error(err: &duckdb::Error) -> bool {
    let msg = err.to_string();
    msg.contains("already exists") || msg.contains("Type with name")
}
